src/BiDirectionalNode.h

```diff
@@ -0,0 +1,44 @@
+//
+// Created by shvmo on 02/05/2020.
+//
+
+#ifndef DATASTRUCTURESWET1_BIDIRECTIONALNODE_H
+#define DATASTRUCTURESWET1_BIDIRECTIONALNODE_H
+
+#include <memory>
+
+using std::shared_ptr;
+using std::weak_ptr;
+
+template<class Type>
+class BiDirectionalNode {
+public:
+    BiDirectionalNode();
+
+    ~BiDirectionalNode();
+
+    Type &getValue();
+
+    shared_ptr<BiDirectionalNode<Type>> getNext() const;
+
+    // Done for optimization when we want to iterate a bit faster
+    BiDirectionalNode<Type> *getNextPointer() const;
```

Comment:
confusing names and perhaps switched by mistake(?)

Reply:
if it is switched then we can just 
rename `getNext()` to `getNextPointer2()`
rename `getNextPointer()` to `getNext()`
and remove the 2 from `getNextPointer2()`

Reply:
ok so after bumping into this a few more times i realized you return a pointer instead of `shared_ptr`.
1. why? is it really that much faster?
2. the getter names are really confusing. 
might want to change `*get<member>Pointer()`
to `*get<member>RawPointer()`
![image](https://user-images.githubusercontent.com/26043145/81995446-1101bd00-9653-11ea-8875-d75159a4bc33.png)

Reply:
Yes it improved speed by 4 times. In some areas. I did it just in case.

Reply:
Not that it matter for efficiency, but just incase the segel does some timeout. Most people had their tests run in 3 seconds, this took 12 seconds. Once I switched some places that were executed a lot to use raw pointers we passed it in less than 3 seconds.

Most people used raw pointers so I guess it's a fair comparison.

Reply:
But I will fix the names

Reply:
awesome. waiting for the changes

Reply:
(in all the classes you added it)